### PR TITLE
Bugfix FXIOS-6749 [v114] Deeplink is called when the browser isn't ready

### PR DESCRIPTION
**Pull request by lmarceau**
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-6749)

### Description
- Ensure we save the Route when we can't navigate to it with coordinators yet
- Separate `BaseCoordinator` into it's own file
- Add logs to ensure we track down the issue if it's not solved
- Add method to know when browser has loaded, then check to navigate to `Route`
- Add check when we finished Launch (survey, onboarding, update views) to navigate to `Route` when those are done showing
- Logger is now part of the `BaseCoordinator`
- Add unit tests + tested manually as well

### Pull requests checks where applicable
- [X] Fill in the three TODOs above (tickets number and description of your work)
- [X] The PR name follows our [naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Accessibility implemented and tested (minimum Dynamic Text and VoiceOver)
- [X] Unit tests written and passing
- [ ] Documentation / comments for complex code and public methods

### CODE REVIEW

File: Client/Coordinators/Browser/BrowserCoordinator.swift [hunk 1]
```diff
@@ -67,6 +68,11 @@ class BrowserCoordinator: BaseCoordinator, LaunchCoordinatorDelegate, BrowserDel
     func didFinishLaunch(from coordinator: LaunchCoordinator) {
         router.dismiss(animated: true, completion: nil)
         remove(child: coordinator)
+
+        // Once launch is done, we check for any saved Route
+        if let savedRoute {
```
Comment (adudenamedruby) [inline Client/Coordinators/Browser/BrowserCoordinator.swift:73]: This look so weird to me. lol

**Comment by adudenamedruby** [comment]
LGTM otherwise.

**Review by OrlaM — APPROVED** [review]

File: Client/Coordinators/Browser/BrowserCoordinator.swift (hunk 1 above)
Comment (lmarceau) [reply Client/Coordinators/Browser/BrowserCoordinator.swift:73]: Yeah I'm trying to get used to it, and be up-to-date with standards. I feel ambivalent to its usage though, we can discuss it in an eng meeting?

**Comment by lmarceau** [comment]
@Mergifyio backport release/v114 release/v115

**Pull request merged**